src/metabase/feature_extraction/async.clj

```diff
@@ -0,0 +1,79 @@
+(ns metabase.feature-extraction.async
+  (:require [metabase.api.common :as api]
+            [metabase.models
+             [computation-job :refer [ComputationJob]]
+             [computation-job-result :refer [ComputationJobResult]]]
+            [toucan.db :as db]))
+
+(defonce ^:private running-jobs (atom {}))
+
+(def ^{:arglists '([job])} done?
+  "Is the computation job done?"
+  (comp some? #{:done :error} :status))
+
+(def ^{:arglists '([job])} running?
+  "Is the computation job still running?"
+  (comp some? #{:running} :status))
+
+(defn- save-result
+  [{:keys [id]} payload]
+  (db/transaction
+    (db/insert! ComputationJobResult
+      :job_id     id
+      :permanence :temporary
+      :payload    payload)
+    (db/update! ComputationJob id :status :done))
+  (swap! running-jobs dissoc id))
+
+(defn- save-error
+  [{:keys [id]} error]
+  (db/transaction
+    (db/insert! ComputationJobResult
+      :job_id     id
+      :permanence :temporary
+      :payload    (Throwable->map error))
+    (db/update! ComputationJob id :status :error))
+  (swap! running-jobs dissoc id))
+
+(defn cancel
+  "Cancel computation job (if still running)."
+  [{:keys [id] :as job}]
+  (when (running? job)
+    (future-cancel (@running-jobs id))
+    (swap! running-jobs dissoc id)
+    (db/update! ComputationJob id :status :canceled)))
+
+(defn compute
+  "Compute closure `f` asynchronously. Returns id of the associated computation
+   job."
+  [f]
+  (let [job (db/insert! ComputationJob
```

Comment:
You should log something to indicate the job started + stopped, and any errors thrown.

Reply:
If the error was in what gets computed, it gets stored same as results (see `save-error`), if it's in the computation framework it 500s same as we do elsewhere. We should catch both Errors and Exceptions though, so things like OutOfMemory below will get caught. Not entirely sure how to handle other Throwables as we do use them for flow control (eg. `api/read-check`).

Reply:
We should still log things in the server logs. It's a basic pattern we use all over the place -- if something on the server goes wrong, throw it into the server logs.

Not all administrators or analysts will have access to that session, or the underlying application database.